src/bun.js/bindings/JSCommonJSExtensions.cpp

```diff
@@ -0,0 +1,181 @@
+    return JSValue::encode({});
+}
+JSC_DEFINE_HOST_FUNCTION(jsLoaderNode, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto scope = DECLARE_THROW_SCOPE(globalObject->vm());
+    throwTypeError(globalObject, scope, "Calling Module._extensions[\".node\"] directly is not implemented."_s);
+    return JSValue::encode({});
+}
+JSC_DEFINE_HOST_FUNCTION(jsLoaderTS, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto scope = DECLARE_THROW_SCOPE(globalObject->vm());
+    throwTypeError(globalObject, scope, "Calling Module._extensions[\".ts\"] directly is not implemented."_s);
+    return JSValue::encode({});
+}
+
+void JSCommonJSExtensions::finishCreation(JSC::VM& vm)
+{
+    Base::finishCreation(vm);
+    ASSERT(inherits(info()));
+
+    Zig::GlobalObject* global = defaultGlobalObject(globalObject());
+    JSC::JSFunction* fnLoadJS = JSC::JSFunction::create(
+        vm,
+        global,
+        2,
+        ""_s,
+        jsLoaderJS,
+        JSC::ImplementationVisibility::Public,
+        JSC::Intrinsic::NoIntrinsic,
+        JSC::callHostFunctionAsConstructor);
+    JSC::JSFunction* fnLoadJSON = JSC::JSFunction::create(
+        vm,
+        global,
+        2,
+        ""_s,
+        jsLoaderJSON,
+        JSC::ImplementationVisibility::Public,
+        JSC::Intrinsic::NoIntrinsic,
+        JSC::callHostFunctionAsConstructor);
+    JSC::JSFunction* fnLoadNode = JSC::JSFunction::create(
+        vm,
+        global,
+        2,
+        ""_s,
+        jsLoaderNode,
+        JSC::ImplementationVisibility::Public,
+        JSC::Intrinsic::NoIntrinsic,
+        JSC::callHostFunctionAsConstructor);
+    JSC::JSFunction* fnLoadTS = JSC::JSFunction::create(
+        vm,
+        global,
+        2,
+        ""_s,
+        jsLoaderTS,
+        JSC::ImplementationVisibility::Public,
+        JSC::Intrinsic::NoIntrinsic,
+        JSC::callHostFunctionAsConstructor);
+
+    this->putDirect(vm, JSC::Identifier::fromString(vm, ".js"_s), fnLoadJS, 0);
+    this->putDirect(vm, JSC::Identifier::fromString(vm, ".json"_s), fnLoadJSON, 0);
+    this->putDirect(vm, JSC::Identifier::fromString(vm, ".node"_s), fnLoadNode, 0);
+    this->putDirect(vm, JSC::Identifier::fromString(vm, ".ts"_s), fnLoadTS, 0);
+    this->putDirect(vm, JSC::Identifier::fromString(vm, ".cts"_s), fnLoadTS, 0);
+    this->putDirect(vm, JSC::Identifier::fromString(vm, ".mjs"_s), fnLoadJS, 0);
+    this->putDirect(vm, JSC::Identifier::fromString(vm, ".mts"_s), fnLoadTS, 0);
+}
+
+extern "C" void NodeModuleModule__onRequireExtensionModify(
+    Zig::GlobalObject* globalObject,
+    const BunString* key,
+    uint32_t kind,
+    JSC::JSValue value);
+
+void onAssign(Zig::GlobalObject* globalObject, JSC::PropertyName propertyName, JSC::JSValue value)
+{
+    if (propertyName.isSymbol()) return;
+    auto* name = propertyName.publicName();
+    if (!name->startsWith("."_s)) return;
+    BunString ext = Bun::toString(name);
+    uint32_t kind = 0;
```

Comment:
this should be an enum

Reply:
i chose not to use an enum because it is used for a single parameter in one very specific function. the enum would have to be manually transcribed anyways, so it doesn't make it safer (until we get translate-c++)